quant: remove incorrect integer cast on histogram observer endpoints

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#83755 quant: remove incorrect integer cast on histogram observer endpoints**

Summary:

In https://github.com/pytorch/pytorch/pull/45630 a cast to int was added
to endpoints of histogram calculation, to make mypy not complain. This is a bug
because we want the histogram endpoints to be in the same dtype as the original
tensor. Reverting this part of the original PR.

Note: this may not have significant numerical impact, as the
`HistogramObserver._non_linear_param_search` function is resilient to
incorrect histogram endpoints and still gives reasonable results for handcrafted
tensors which are likely to have integer casts of histogram endpoints
give incorrect histograms. Because of this, I did not add an additional test case
to save time, and the main benefit will be code clarity.  If the
`_non_linear_param_search` function changes in the future, this may need to be
revisited.

Test plan:

```
python test/test_quantization.py -k histogram
```

vkuzo committed:
quant: remove incorrect integer cast on histogram observer endpoints

Summary:

In https://github.com/pytorch/pytorch/pull/45630 a cast to int was added
to endpoints of histogram calculation, to make mypy not complain. This is a bug
because we want the histogram endpoints to be in the same dtype as the original
tensor. Reverting this part of the original PR.

Note: this may not have significant numerical impact, as the
`HistogramObserver._non_linear_param_search` function is resilient to
incorrect histogram endpoints and still gives reasonable results for handcrafted
tensors which are likely to have integer casts of histogram endpoints
give incorrect histograms. Because of this, I did not add an additional test case
to save time, and the main benefit will be code clarity.  If the
`_non_linear_param_search` function changes in the future, this may need to be
revisited.

Test plan:

```
python test/test_quantization.py -k histogram
```

[ghstack-poisoned]

vkuzo committed:
Update on "quant: remove incorrect integer cast on histogram observer endpoints"


Summary:

In https://github.com/pytorch/pytorch/pull/45630 a cast to int was added
to endpoints of histogram calculation, to make mypy not complain. This is a bug
because we want the histogram endpoints to be in the same dtype as the original
tensor. Reverting this part of the original PR.

Note: this may not have significant numerical impact, as the
`HistogramObserver._non_linear_param_search` function is resilient to
incorrect histogram endpoints and still gives reasonable results for handcrafted
tensors which are likely to have integer casts of histogram endpoints
give incorrect histograms. Because of this, I did not add an additional test case
to save time, and the main benefit will be code clarity.  If the
`_non_linear_param_search` function changes in the future, this may need to be
revisited.

Test plan:

```
python test/test_quantization.py -k histogram
```

[ghstack-poisoned]

vkuzo committed:
Update on "quant: remove incorrect integer cast on histogram observer endpoints"


Summary:

In https://github.com/pytorch/pytorch/pull/45630 a cast to int was added
to endpoints of histogram calculation, to make mypy not complain. This is a bug
because we want the histogram endpoints to be in the same dtype as the original
tensor. Reverting this part of the original PR.

Note: this may not have significant numerical impact, as the
`HistogramObserver._non_linear_param_search` function is resilient to
incorrect histogram endpoints and still gives reasonable results for handcrafted
tensors which are likely to have integer casts of histogram endpoints
give incorrect histograms. Because of this, I did not add an additional test case
to save time, and the main benefit will be code clarity.  If the
`_non_linear_param_search` function changes in the future, this may need to be
revisited.

Test plan:

```
python test/test_quantization.py -k histogram
```

[ghstack-poisoned]

jerryzh168 reviewed (approved):
do we need to run any numerical test for internal models?

jerryzh168 commented:
also CI errors looks related

vkuzo commented:
yeah the CI failure is real, I need to debug that first before landing this.

PyPER does not use HistogramObserver, we will see internal CI for other ones - need to make OSS CI green first.

vkuzo committed:
Update on "quant: remove incorrect integer cast on histogram observer endpoints"


Summary:

In https://github.com/pytorch/pytorch/pull/45630 a cast to int was added
to endpoints of histogram calculation, to make mypy not complain. This is a bug
because we want the histogram endpoints to be in the same dtype as the original
tensor. Reverting this part of the original PR.

Note: this may not have significant numerical impact, as the
`HistogramObserver._non_linear_param_search` function is resilient to
incorrect histogram endpoints and still gives reasonable results for handcrafted
tensors which are likely to have integer casts of histogram endpoints
give incorrect histograms. Because of this, I did not add an additional test case
to save time, and the main benefit will be code clarity.  If the
`_non_linear_param_search` function changes in the future, this may need to be
revisited.

Test plan:

```
python test/test_quantization.py -k histogram
```

[ghstack-poisoned]
